data/gitg.1
.TH GITG "1" "December 2018" "gitg" "User Commands"
.SH NAME
gitg \- Git repository viewer
.SH SYNOPSIS
.B gitg
[\fI\,OPTION\/\fR]... [\fI\,REPOSITORY\/\fR]...
.SH DESCRIPTION
.B gitg
is a graphical user interface for git. It aims at being a small,
fast and convenient tool to visualize the history of git repositories.
Besides visualization, gitg also provides several utilities to manage your
repository and commit your work.
.SH OPTIONS
Help Options:
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help options
.TP
\fB\-\-help\-all\fR
Show all help options
.TP
\fB\-\-help\-gtk\fR
Show GTK+ Options
.PP
Application Options:
.TP
\fB\-v\fR, \fB\-\-version\fR
Show the application?s version
.TP
\fB\-\-activity\fR
Start gitg with a particular activity
.TP
\fB\-c\fR, \fB\-\-commit\fR
Start gitg with the commit activity (shorthand for \fB\-\-activity\fR commit)
.TP
\fB\-\-no\-wd\fR
Do not try to load a repository from the current working directory
.TP
\fB\-\-standalone\fR
Run gitg in standalone mode
.TP
\fB\-\-display\fR=\fI\,DISPLAY\/\fR
X display to use
.TP
\fB\-a\fR, \fB\-\-all\fR
Select all commits by default in the history activity
.TP
\fB\-b\fR, \fB\-\-branches\fR
Select all branches by default in the history activity
.TP
\fB\-r\fR, \fB\-\-remotes\fR
Select all remotes by default in the history activity
.TP
\fB\-t\fR, \fB\-\-tags\fR
Select all tags by default in the history activity
.TP
\fB\-s\fR, \fB\-\-select\-reference\fR=\fI\,REFERENCE\/\fR
Select the specified reference by default in the history activity
.PP
.B gitg
will try to load any REPOSITORY passed as command line argument (multiple paths can be passed).
.PP
If no repository is specified,
.B gitg
will try to open the repository specified by the \fBGIT_DIR\fR environment variable;
if \fBGIT_DIR\fR is not set
.B gitg
will try to open the repository at the current working directory.
.SH ENVIRONMENT VARIABLES
\fBGIT_DIR\fR
.RS 4
If the
\fBGIT_DIR\fR
environment variable is set, and no repositories have been passed as command line arguments, then it specifies a path to use instead of the default
\fB\&.git\fR
for the base of the repository\&.
.RE
.SH AUTHORS
Jesse van den Kieboom  <jesse@icecrew.nl>
.br
Alberto Fanjul  <albfan@gnome.org>


NEWS
= gitg 3.30.1 =
Version 3.30.1 was released on 2018-10-25

== Bugs fixed ==
- Meson improvements
- Fix parsing title and body for message commit
- Search repositories ignoring case
- Added standalone mode
- Allow to config patience algorithm
- GUI improvements
- Removed autotools buildsystem
- Fixed mac compilation
- Misc fixes

= gitg 3.30.0 =
Version 3.30.0 was released on 2018-10-07

== Bugs fixed ==
- Detect dark theme from its background color
- Buildsystem meson added
- Added flatpak deployments
- Updated translations
- Misc fixes

= gitg 3.26.0 =
Version 3.26.0 was released on 2017-08-09

== Bugs fixed ==
- Updated translations
- Misc fixes

= gitg 3.24.0 =
Version 3.24.0 was released on 2017-04-30

== Bugs fixed ==
- Updated translations
- Misc fixes

= gitg 3.23.90 =
Version 3.23.90 was released on 2017-02-16

== Changes ==
- Fix for libgit2-glib api change

== Bugs fixed ==
- Do not duplicate the fds

= gitg 3.23.0 =
Version 3.23.0 was released on 2016-11-26

== Changes ==
- Bump libgit2-glib version to 0.24.4
- Show dash activity menu if shell does not show app menu
- Add diff renderer for images

== Bugs fixed ==
- history: Fix printf format strings
- Use same callbacks instance during remote connection life cycle
- win32 installer fixes

= gitg 3.21.91 =
Version 3.21.91 was released on 2016-08-27

== Changes ==
- Add context menu to open files from the diff
- Better support for bare and empty repositories
- Add support for ssh short form remote urls
- Improved selection toggling for incremental staging
- Disable use of gravatar by default for privacy

== Bugs fixed ==
- Set primary selection to selected commit SHA1
- Scrollable commit message in diff view
- Incremental loading on large repositories
- Faster switching of large repositories
- Create global git config if it does not exist yet
- Remember opened projects regardless of gnome privacy settings

= gitg 3.20.0 =
Version 3.20.0 was released on 2016-03-20

== Bugs fixed ==
- Updated translations
- Misc fixes

= gitg 3.19.6 =
Version 3.19.6 was released on 2016-02-04

== Changes ==
- Make use of shortcut window optional

== Bugs fixed ==
- Make it build again on osx and windows

= gitg 3.19.5 =
Version 3.19.5 was released on 2016-01-26

== Changes ==
- Improved Windows installer graphics
- Add context menu on Windows
- Provide shortcuts window
- Rework dash view

== Bugs fixed ==
- Do not hang on highlighting commit message

= gitg 3.19.4 =
Version 3.19.4 was released on 2016-01-08

== Changes ==
- Add OSX bundle generator
- Add syntax highlighting support on diffs
- Show untracked content on stage

== Bugs fixed ==
- Fix OSX paths
- Missing includes

= gitg 3.19.3 =
Version 3.19.3 was released on 2015-12-23

== Changes ==
- Add script to create windows installer
- Add support for commit.template config setting
- Dim default avatar image

== Bugs fixed ==
- Add icon to the windows binary
- Handle console on windows


= gitg 3.19.2 =
Version 3.19.2 was released on 2015-12-22

== Changes ==
- Support for Windows and OSX
- Add platform specific styles

== Bugs fixed ==
- Fix dark style of diff-stat widget
- Fix selectable button visibilities


= gitg 3.19.1 =
Version 3.19.1 was released on 2015-12-21

== Changes ==
- Rework diff view to use GtkSourceView instead of Webkit
- Initial support of dark theme
- Better support on OSX
- Improved commit selection


= gitg 3.17.1 =
Version 3.17.1 was released on 2015-08-08

== Changes ==
- Make use of gravatar service optional (preference)
- Stage/unstage all selected items when pressing Enter
- Add command line arguments to modify default history
  activity selection (--all, --branches, --remotes, --tags)
- Add deleting of untracked files
- Improvements to the dash/overview UX
- Improvements to collapse/uncollapse behavior of diff view


== Bugs fixed ==
- Get icons at correct size in diff view
- Small UI polish
- Fixed authentication for cloning
- Only use animations of gtk_enable_animations is set
- Fix collation errors for python diffs
- Fix race conditions in repository loading


= gitg 3.16.1 =
Version 3.16.1 was released on 2015-04-29

== Changes ==
  * Include passed in cflags for threading support check
  * Updated translations

= gitg 3.16.0 =
Version 3.16.0 was released on 2015-03-24

== Changes ==
  * Updated translations

= gitg 3.15.2 =
Version 3.15.2 was released on 2015-03-17

== Changes ==
  * Show committer date in commit info
  * Deep extend diff settings to preserve defaults
  * Show committer in diff when differs from author
  * Select row when going into selection mode
  * Do not use null repository in remote manager
  * Provide a symbolic variant of the app icon
  * Make links in commit messages clickable
  * Add gitg config to ignore certain submodules
  * Updated translations

= gitg 3.15.1 =
Version 3.15.1 was released on 2015-02-09

== Changes ==
  * Fix building with vala 0.25
  * Bump libgit2-glib to 0.22.0
  * Implement basic history search
  * Implement basic fetch
  * Implement remote state tracking
  * Add remote management
  * Show hunk context in diff
  * Implement configuring of mainlines
  * Implement preserving mainlines on history lanes
  * Implement opening file from staging area
  * Allow choosing merge diff parent
  * Implement submodule patch stage/unstage
  * Implement workdir submodule stage/unstage
  * Show submodules in dash
  * Updated translations

= gitg 3.14.0 =
Version 3.14.0 was released on 2014-09-24

== Bug fixes ==
  * Fix building with vala 0.25

== Updated translations ==
  * update zh_CN translation: YunQiang Su <yqsu@src.gnome.org>
  * German: Benjamin Steinwender <b@stbe.at>
  * Slovak: Peter Vágner <pvdeejay@gmail.com>
  * Russian: Yuri Myasoedov <ymyasoedov@yandex.ru>
  * Serbian: Мирослав Николић <miroslavnikolic@rocketmail.com>
  * Swedish: Sebastian Rasmussen <sebras@gmail.com>
  * Danish: Kenneth Nielsen <k.nielsen81@gmail.com>
  * Finnish translation update by Jiri Grönroos: Ville-Pekka Vainio <vpvainio@iki.fi>

gitg/gitg.c
/*
 * gitg.c
 * This file is part of gitg - git repository viewer
 *
 * Copyright (C) 2009 - Jesse van den Kieboom
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <gtk/gtk.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <stdlib.h>
#include <string.h>
#include <gtksourceview/gtksource.h>
#include <libgitg/gitg-debug.h>

#include "gitg-window.h"
#include "config.h"
#include "gitg-dirs.h"
#include "gitg-utils.h"
#include "gitg-uri.h"

static gboolean commit_mode = FALSE;
static gchar *select_sha1 = NULL;

static void
show_version_and_quit (void)
{
	g_print ("%s - Version %s\n", g_get_application_name (), VERSION);

	exit (0);
}

static GOptionEntry entries[] = 
{
	{ "version", 'V', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK,
	  show_version_and_quit, N_("Show the application's version"), NULL },
	{ "commit", 'c', 0, G_OPTION_ARG_NONE, &commit_mode, N_("Start gitg in commit mode") },
	{ "select", 's', 0, G_OPTION_ARG_STRING, &select_sha1, N_("Select commit after loading the repository") },
	{ NULL }
};

static void
parse_options (int *argc, char ***argv)
{
	GError *error = NULL;
	GOptionContext *context;

	context = g_option_context_new (_("- git repository viewer"));

	// Ignore unknown options so we can pass them to git
	g_option_context_set_ignore_unknown_options (context, TRUE);
	g_option_context_add_main_entries (context, entries, GETTEXT_PACKAGE);
	g_option_context_add_group (context, gtk_get_option_group (TRUE));

	if (!g_option_context_parse (context, argc, argv, &error))
	{
		g_print ("option parsing failed: %s\n", error->message);
		g_error_free (error);
		exit (1);
	}

	g_option_context_free(context);
}

static gboolean
on_window_delete_event (GtkWidget *widget, gpointer userdata)
{
	gtk_main_quit ();
	return FALSE;
}

static GitgWindow *
build_ui (void)
{
	GtkBuilder *builder = gitg_utils_new_builder ("gitg-window.ui");

	GtkWidget *window = GTK_WIDGET (gtk_builder_get_object(builder, "window"));
	gtk_widget_show_all (window);

	g_signal_connect_after (window, "destroy", G_CALLBACK (on_window_delete_event), NULL);
	g_object_unref (builder);

	return GITG_WINDOW (window);
}

static void
set_language_search_path (void)
{
	GtkSourceLanguageManager *manager = gtk_source_language_manager_get_default ();
	gchar const * const *orig = gtk_source_language_manager_get_search_path (manager);
	gchar const **dirs = g_new0 (gchar const *, g_strv_length ((gchar **)orig) + 2);
	guint i = 0;

	while (orig[i])
	{
		dirs[i + 1] = orig[i];
		++i;
	}

	gchar *path = gitg_dirs_get_data_filename ("language-specs", NULL);
	dirs[0] = path;
	gtk_source_language_manager_set_search_path (manager, (gchar **)dirs);
	g_free (path);

	g_free (dirs);
}

static void
set_style_scheme_search_path (void)
{
	GtkSourceStyleSchemeManager *manager = gtk_source_style_scheme_manager_get_default ();

	gchar *path = gitg_dirs_get_data_filename ("styles", NULL);
	gtk_source_style_scheme_manager_prepend_search_path (manager, path);
	g_free (path);
}

static void
set_icons (void)
{
	static gchar const *icon_infos[] = {
		"gitg16x16.png",
		"gitg22x22.png",
		"gitg24x24.png",
		"gitg32x32.png",
		"gitg48x48.png",
		"gitg64x64.png",
		"gitg128x128.png",
		NULL
	};

	int i;
	GList *icons = NULL;

	for (i = 0; icon_infos[i]; ++i)
	{
		gchar *filename = gitg_dirs_get_data_filename ("icons", icon_infos[i], NULL);
		GdkPixbuf *pixbuf = gdk_pixbuf_new_from_file (filename, NULL);
		g_free (filename);

		if (pixbuf)
		{
			icons = g_list_prepend (icons, pixbuf);
		}
	}

	gtk_window_set_default_icon_list (icons);

	g_list_free_full (icons, g_object_unref);
}

int
main (int argc, char **argv)
{
	gboolean ret;

	gitg_debug_init ();

	bindtextdomain (GETTEXT_PACKAGE, GITG_LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	g_set_prgname ("gitg");

	/* Translators: this is the application name as in g_set_application_name */
	g_set_application_name (_("gitg"));

	gitg_dirs_initialize (argc, argv);
	gtk_init (&argc, &argv);
	parse_options (&argc, &argv);

	set_language_search_path ();
	set_style_scheme_search_path ();
	set_icons ();

	GitgWindow *window = build_ui ();

	ret = gitg_window_load_repository_for_command_line (window,
	                                                    argc - 1,
	                                                    (gchar const **)argv + 1,
	                                                    select_sha1);

	if (commit_mode && ret)
	{
		gitg_window_show_commit (window);
	}

	gtk_main ();

	return 0;
}
